### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by alexander-veit**
This PR introduces various bug fixes and new features.

- Added small example files to check non-standard reads (deletions, soft/hard clipping)
- Build out `es` modules when "npm run build" is executed
- Add option `workerScriptLocation`. This is needed when the worker script is not in the same folder as the pileup script
- The client side data fetcher now accepts `bamUrl` and `baiUrl`. `url` is still supported
- Added options `plusStrandColor` and `minusStrandColor`. When set, the reads are colored according to their strand.
- Added the option `showReadCounts` and `readCountHeight,` which displays read count information (like a bar track) on top of the track
- Improved mouseovers for reads and read counts
- Bumped version of GMOD/bam-js, because the current version did not allow to retrieve the read sequence
- Variants are now correctly colored based on the read sequence.
- Added visualization of reads that contain hard clipping
- Fixed a bug where MD strings would not be correctly parsed, when a deletion is present, e.g., `85^A16`
- Added stripes to represent deletions. It's not extremely pretty, but it's easier to distinguish. Please let me know if you don't want that. There is an example in the `index.html`, that you can have a look at.
- When there is soft clipping at the beginning of a read, substitutions that are following are now correctly extracted
- Reads (not substitutions) have been shifted by one. It now lines up with the sequence track and the bam file. See below. Please check.

Screen Shot 2020-11-06 at 1 21 14 PM
Screen Shot 2020-11-02 at 12 01 04 PM

Read count information:
Screen Shot 2020-11-06 at 1 25 15 PM

- [ ] Update the version number in unpkg link in README.md if building a new release

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "higlass-pileup",
-  "version": "0.4.0",
+  "version": "0.4.1",
```
Comment (pkerpedjiev) [inline package.json:3]: This should definitely be at least a minor version bump if not a major. We could just go ahead and release v1.0.0 with these changes.

File: src/PileupTrack.js [hunk 2]
```diff
@@ -188,9 +187,14 @@ const PileupTrack = (HGC, ...args) => {
 
   class PileupTrackClass extends HGC.tracks.Tiled1DPixiTrack {
     constructor(context, options) {
+      let baseUrl = `${getThisScriptLocation()}/`;
+      if (options.workerScriptLocation) {
```
Comment (pkerpedjiev) [inline src/PileupTrack.js:191]: Could you add this option to the README?

File: src/PileupTrack.js [hunk 3]
```diff
@@ -343,27 +361,27 @@ varying vec4 vColor;
       }
 
       const fetchedTileKeys = Object.keys(this.fetchedTiles);
-      fetchedTileKeys.forEach((x) => {
+      fetchedTileKeys.forEach(x => {
```
Comment (pkerpedjiev) [inline src/PileupTrack.js:364]: Are you using an older version of prettier? Parenthesis around anonymous function arguments are on by default as of prettier v2.0

File: src/PileupTrack.js [hunk 4]
```diff
@@ -492,22 +511,29 @@ varying vec4 vColor;
       //   .domain([0, this.prevRows.length])
       //   .range([0, this.dimensions[1]]);
       // HGC.utils.trackUtils.drawAxis(this, valueScale);
-      this.trackNotFoundText.text = 'Pete rules11!';
+      this.trackNotFoundText.text = 'Pete rules!!!';
```
Comment (pkerpedjiev) [inline src/PileupTrack.js:514]: Ummm... we should probably change this 🤣

File: src/PileupTrack.js [hunk 5]
```diff
@@ -562,26 +588,54 @@ varying vec4 vColor;
...
                   // + `CIGAR: ${read.cigar || ''} MD: ${read.md || ''}`);
                 }
               }
             }
           }
         }
+
         // var val = self.yScale.domain()[index];
+        if (
+          this.options.showReadCounts &&
```
Comment (pkerpedjiev) [inline src/PileupTrack.js:613]: I think what you refer to read counts is usually referred to as "coverage".

File: src/bam-fetcher-worker.js [hunk 6]
```diff
@@ -848,31 +952,50 @@ const renderSegments = (
         xLeft = from;
         xRight = to;
 
-        addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.BG);
-
-        const substitutions = getSubstitutions(segment);
+        //console.log(xLeft, yTop, xRight - xLeft, height)
```
Comment (pkerpedjiev) [inline src/bam-fetcher-worker.js:955]:
```suggestion
```

File: src/bam-fetcher-worker.js [hunk 7]
```diff
@@ -848,31 +952,50 @@ const renderSegments = (
...
+        if(segment.strand === '+' && trackOptions.plusStrandColor){
+          addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.PLUS_STRAND);
+        }
+        else if(segment.strand === '-' && trackOptions.minusStrandColor){
+          addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.MINUS_STRAND);
+        }
+        else{
+          addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.BG);
+        }
+        //console.log(allColors)
```
Comment (pkerpedjiev) [inline src/bam-fetcher-worker.js:965]:
```suggestion
```

File: src/bam-utils.js [hunk 8]
```diff
@@ -20,46 +24,74 @@ Object.keys(PILEUP_COLORS).map((x, i) => {
...
+          bamSeqShift: bamSeqShift,
         });
       }
 
       currNum = 0;
       currPos += 1;
     }
   }
-
+  // if(!useCounts){
```
Comment (pkerpedjiev) [inline src/bam-utils.js:84]: Might as well delete this, no?

File: src/bam-utils.js [hunk 9]
```diff
@@ -20,46 +24,74 @@ Object.keys(PILEUP_COLORS).map((x, i) => {
...
+  //   console.log(mdString, substitutions)
+  // }
+  
   return substitutions;
 };
 
-export const getSubstitutions = (segment) => {
+export const getSubstitutions = (segment, seq) => {
```
Comment (pkerpedjiev) [inline src/bam-utils.js:91]: What is seq? I've set a bad precedent here but docstrings would be good to have.

File: src/bam-utils.js [hunk 10]
```diff
@@ -82,29 +113,28 @@ export const getSubstitutions = (segment) => {
...
-          sub,
           type: 'D',
         });
         currPos += sub.length;
       } else if (sub.type === 'N') {
         substitutions.push({
           pos: currPos,
           length: sub.length,
-          sub,
+          //sub,
```
Comment (pkerpedjiev) [inline src/bam-utils.js:129]:
```suggestion
```

File: src/bam-utils.js [hunk 11]
```diff
@@ -82,29 +113,28 @@ export const getSubstitutions = (segment) => {
...
           type: 'N',
         });
         currPos += sub.length;
       } else if (sub.type === '=' || sub.type === 'M') {
         currPos += sub.length;
       } else {
         // console.log('skipping:', sub.type);
+        // console.log(cigarSubs)
```
Comment (pkerpedjiev) [inline src/bam-utils.js:137]:
```suggestion
```

I know, I'm guilty as well but we should probably strive to get rid of useless console log statements.

**Review by pkerpedjiev — APPROVED** [review]
This is awesome! Thanks a ton for submitting this.

I added a bunch of minor comments but overall it looks great! Do you think there's any way to benchmark how fast it renders for a large-ish file and compare before these changes and after?

**Review by pkerpedjiev — COMMENTED** [review]
Tried to change to comment rather than approve but apparently it doesn't work. I just wanted the review status to be set to `comment` to ensure that the little things get taken care of before merging.

**Comment by alexander-veit** [comment]
Hey Pete, thanks a lot for the detailed review and the comments. I am out until Thursday, but will get started on the changes and the benchmark then right away.

File: package.json (hunk 1 above)
Comment (alexander-veit) [reply package.json]: Done

File: src/PileupTrack.js (hunk 2 above)
Comment (alexander-veit) [reply src/PileupTrack.js:191]: Done

File: src/PileupTrack.js (hunk 4 above)
Comment (alexander-veit) [reply src/PileupTrack.js]: I reluctantly changed it ;-)

File: src/bam-utils.js (hunk 9 above)
Comment (alexander-veit) [reply src/bam-utils.js:94]: Done

File: src/bam-utils.js (hunk 11 above)
Comment (alexander-veit) [reply src/bam-utils.js]: Agreed.

File: src/PileupTrack.js (hunk 5 above)
Comment (alexander-veit) [reply src/PileupTrack.js]: I changed it to coverage everywhere

File: src/PileupTrack.js (hunk 3 above)
Comment (alexander-veit) [reply src/PileupTrack.js]: Thanks for noticing. I am using the correct version now.

**Comment by alexander-veit** [comment]
I modified it according to your comments.

Furthermore, I added a new `segmentsToRow` function. It is a bit easier to read in my opinion, and resolves the 'overlapping reads' issue. Performance-wise it seems a bit faster than the old function, when you initially load the display (i.e. `prevSegments` is empty). I tested it with a window size of 150kb and 50,000 segments. When you zoom around it is  equal or a little bit slower than the old function (because of the way `prevSegments` is processed. However, in practice I could not notice any difference. If you experience this differently, please let me know.

Sorry, that I included it in this PR. It uses some of the modifications already made and it was easier that way. It should be easy to review, it is just an altered `segmentsToRow` function.

The coverage 'track' is currently a bit slow (it is switched off by default). That's not an issue when the window size is reasonable, but when you look at a 150kb window it takes around 80% of the `renderSegments` time. This is of course too much. If it's ok for you, I would like to address this in the next PR.

**Comment by alexander-veit** [comment]
I have now a much faster version of the coverage track. I will create a PR for that once this PR is merged.

File: src/index.html [hunk 12]
```diff
@@ -66,30 +67,127 @@
...
+            "data": {
+              "type": "fasta",
+              "faiUrl": "https://aveit.s3.amazonaws.com/higlass/data/sequence/hg38.fa.fai",
+              "fastaUrl": "https://aveit.s3.amazonaws.com/higlass/data/sequence/hg38.fa",
+              "chromSizesUrl": "https://aveit.s3.amazonaws.com/higlass/data/sequence/hg38.mod.chrom.sizes"
+            },
+            "type": "horizontal-sequence",
+            "width": 768,
+            "height": 25,
+            "server": "https://cgap-higlass.com/api/v1",
```
Comment (pkerpedjiev) [inline src/index.html:144]: Not related to this PR, but why does the sequence track require a server if urls are provided in the `data` section?

File: src/index.html [hunk 13]
```diff
@@ -98,8 +196,38 @@
                 "#DCDCDC"
               ]
             },
-            "width": 651,
-            "height": 362
+            "height": 500,
+            "uid": "FylkvVBTSumoJ959HT4-5B",
+            "data": {
+              "type": "bam",
+              // Example Pete
```
Comment (pkerpedjiev) [inline src/index.html:203]: You can take this out.

File: src/index.html [hunk 14]
```diff
@@ -66,30 +67,127 @@
   "exportViewUrl": "/api/v1/viewconfs",
   "views": [
     {
-      "initialXDomain": [
-        1069795.099783097,
-        1070508.6726894341
-      ],
+      // Example Pete
```
Comment (pkerpedjiev) [inline src/index.html:70]: You can take this out :-)

**Review by pkerpedjiev — APPROVED** [review]
Looks great! Thanks a ton for all these updates 🙏

File: src/index.html (hunk 12 above)
Comment (alexander-veit) [reply src/index.html]: I doesn't need it. I removed it.

**Pull request merged**